// typedoc-plugin-merge-modules/dist/merger/module_bundle.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ModuleBundle = void 0;
const utils_1 = require("../utils");
/**
 * Name of the comment tag that can be used to mark a module as the target module within the bundle.
 * The target module is the one into which all the modules of the bundle are merged.
 */
const targetModuleCommentTag = "@mergeTarget";
/**
 * Class representing a group of modules.
 */
class ModuleBundle {
    /**
     * Creates a new module bundle instance.
     * @param project The project in which all modules are in.
     */
    constructor(project) {
        /** The modules of the bundle. */
        this.modules = new Array();
        this.project = project;
    }
    /**
     * Adds a module to the bundle.
     * @param module The module to add.
     */
    add(module) {
        this.modules.push(module);
    }
    /**
     * Merges the modules of the bundle into one module.
     */
    merge() {
        const childrenOfAllModules = this.modules
            .map((m) => m.children)
            .filter((m) => m !== undefined)
            .reduce((acc, val) => acc.concat(val), []);
        // get target module
        const targetModule = this.getTargetModule();
        (0, utils_1.removeTagFromCommentsOf)(targetModule, targetModuleCommentTag);
        // set target module for all children
        childrenOfAllModules.forEach((child) => (child.parent = targetModule));
        targetModule.children = childrenOfAllModules;
        // remove rest modules
        this.modules.forEach((module) => {
            if (module !== targetModule) {
                delete module.children;
                this.project.removeReflection(module);
            }
        });
    }
    /**
     * Returns the module from the bundle that should be used as the target module.
     * @returns The target module.
     */
    getTargetModule() {
        // 1. search for the first module which is marked with a specific tag
        const firstModuleWithTargetTag = this.modules.find((module) => module.comment && module.comment.blockTags.findIndex((ct) => ct.tag === targetModuleCommentTag) !== -1);
        if (firstModuleWithTargetTag) {
            return firstModuleWithTargetTag;
        }
        // 2. search for the first module with a comment
        const firstModuleWithComment = this.modules.find((module) => { var _a, _b; return ((_b = (_a = module.comment) === null || _a === void 0 ? void 0 : _a.summary.length) !== null && _b !== void 0 ? _b : 0) > 0; });
        if (firstModuleWithComment) {
            return firstModuleWithComment;
        }
        // 3. default: pick the first module
        return this.modules[0];
    }
}
exports.ModuleBundle = ModuleBundle;
