// typedoc-plugin-merge-modules/dist/plugin_options.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PluginOptions = void 0;
const typedoc_1 = require("typedoc");
/**
 * Class storing the options of the plugin.
 */
class PluginOptions {
    constructor() {
        /** Defines if the plugin should rename default exports to their original name. */
        this._renameDefaults = true;
        /** Defines how the plugin should merge modules. */
        this._mode = "project";
    }
    /**
     * Returns if the plugin should rename default exports to their original name.
     * @returns True, if the plugin should rename default exports to their original name, otherwise false.
     */
    get renameDefaults() {
        return this._renameDefaults;
    }
    /**
     * Returns how the plugin should merge modules.
     * @returns How the plugin should merge modules.
     */
    get mode() {
        return this._mode;
    }
    /**
     * Adds the command line options of the plugin to the TypeDoc application.
     * @param typedoc The TypeDoc application.
     */
    // eslint-disable-next-line class-methods-use-this
    addToApplication(typedoc) {
        typedoc.options.addDeclaration({
            type: typedoc_1.ParameterType.Boolean,
            name: "mergeModulesRenameDefaults",
            help: "If true default exports are renamed to their original name as their module is merged.",
            defaultValue: true,
        });
        typedoc.options.addDeclaration({
            type: typedoc_1.ParameterType.Map,
            map: new Map([
                ["project", "project"],
                ["module", "module"],
                ["module-category", "module-category"],
                ["off", "off"],
            ]),
            name: "mergeModulesMergeMode",
            help: "Defines how the plugin should merge modules.",
            defaultValue: "project",
        });
    }
    /**
     * Reads the values of the plugin options from the application options.
     * @param typedoc The TypeDoc application.
     */
    readValuesFromApplication(typedoc) {
        this._renameDefaults = typedoc.options.getValue("mergeModulesRenameDefaults");
        this._mode = typedoc.options.getValue("mergeModulesMergeMode");
    }
}
exports.PluginOptions = PluginOptions;
